dep_tools/task.py

```
@@ -0,0 +1,83 @@
+        self.writer = writer
+        self.logger = logger
+
+    @abstractmethod
+    def run(self):
+        pass
+
+
+class AreaTask(Task):
+    def __init__(
+        self,
+        id: str,
+        area: GeoDataFrame,
+        loader: Loader,
+        processor: Processor,
+        writer: Writer,
+        logger: Logger = getLogger(),
+        send_area_to_processor: bool = True,
+    ):
+        super().__init__(loader, processor, writer, logger)
+        self.area = area
+        self.send_area_to_processor = send_area_to_processor
+        self.id = id
+
+    def run(self):
+        input_data = self.loader.load(self.area)
+
+        processor_kwargs = (
+            dict(area=self.area) if self.processor.send_area_to_processor else dict()
+        )
+        output_data = self.processor.process(input_data, **processor_kwargs)
+        self.writer.write(output_data, self.id)
+
+
+class LoggingAreaTask(AreaTask):
+    def run(self):
+        try:
+            input_data = self.loader.load(self.area)
+        except EmptyCollectionError as e:
+            self.logger.debug([self.id, "no items for areas"])
+            raise e
+
+        except Exception as e:
+            self.logger.debug([self.id, "load error", e])
+            raise e
+
+        processor_kwargs = (
+            dict(area=self.area) if self.processor.send_area_to_processor else dict()
+        )
+        try:
+            output_data = self.processor.process(input_data, **processor_kwargs)
+        except Exception as e:
+            self.logger.debug([self.id, "processor error", e])
+            raise e
+
+        if output_data is None:
+            self.logger.debug([self.id, "no output from processor"])
+            raise NoOutputError()
+        try:
+            paths = self.writer.write(output_data, self.id)
+        except Exception as e:
+            self.logger.error([self.id, "error", "", e])
+            raise e
+
+        self.logger.debug([self.id, "complete", paths])
```

I think this whole section should be a single Try/Except.

So it goes:

```python
try:
    self.loader.load()
    self.processor.process()
    self.writer.write()
except Exception as e:
    log(f"Tasked failed with exception: {e}
    exit(1)
```

Another option is to not even handle exceptions and just let them kill the app... it's not as clean though.

Point is that the exceptions should be descriptive, so then we don't need to double handle like "caught THING: log(THING happened)"

---

To your second point, that' s why I made the "naked"  AreaTask...just let them bubble up. 

To the first, I wanted to categorize the error types when logging to csv in case the error wasn't clear from the exception type. I guess I'd say don't use that one if it doesn't fit your needs. I get your point though!

---

Oh, `getLogger` is getting an Azure CSV logger.

Maybe we should be clear that it's an AzureLogger, and not a standard logger too.